google/cloud/firestore_bundle/bundle.py

```diff
@@ -0,0 +1,439 @@
+                    queries=original_queries,
+                ),
+            )
+
+        if query_name:
+            bundled_document = self.documents.get(full_document_path)
+            bundled_document.metadata.queries.append(query_name)  # type: ignore
+
+        self._update_last_read_time(snapshot.read_time)
+        # Flush this if it was cached
+        self._deserialized_metadata = None
+        return self
+
+    def add_named_query(self, name: str, query: BaseQuery) -> "FirestoreBundle":
+        """Adds a query to the bundle, referenced by the provided name.
+
+        Args:
+            name (str): The name by which the provided query should be referenced.
+            query (Query): Query of documents to be fully loaded and stored in
+                the bundle for future access.
+
+        Example:
+
+            from google.cloud import firestore
+
+            db = firestore.Client()
+            collection_ref = db.collection(u'users')
+
+            bundle = firestore.FirestoreBundle('my bundle')
+            bundle.add_named_query('all the users', collection_ref._query())
+
+        Returns:
+            FirestoreBundle: self
+
+        Raises:
+            ValueError: If anything other than a BaseQuery (e.g., a Collection)
+                is supplied. If you have a Collection, call its `_query()`
+                method to get what this method expects.
+            ValueError: If the supplied name has already been added.
+        """
+        if not isinstance(query, BaseQuery):
+            raise ValueError(
+                "Attempted to add named query of type: "
+                f"{type(query).__name__}. Expected BaseQuery.",
+            )
+
+        if self.named_queries.get(name):
+            raise ValueError(f"Query name conflict: {name} has already been added.")
+
+        # Execute the query and save each resulting document
+        _read_time = self._save_documents_from_query(name, query)
+
+        # Actually save the query to our local object cache
+        self._save_named_query(name, query, _read_time)
+        # Flush this if it was cached
+        self._deserialized_metadata = None
+        return self
+
+    def _save_documents_from_query(self, name, query: BaseQuery) -> datetime.datetime:
+        _read_time = datetime.datetime.min.replace(tzinfo=UTC)
+        if isinstance(query, AsyncQuery):
+            import asyncio
+
+            loop = asyncio.get_event_loop()
+            return loop.run_until_complete(self._process_async_query(name, query))
+
+        # `query` is now known to be a non-async `BaseQuery`
+        doc: DocumentSnapshot
+        for doc in query.stream():  # type: ignore
+            self.add_document(doc, query_name=name)
+            _read_time = doc.read_time
+        return _read_time
+
+    def _save_named_query(
+        self, name: str, query: BaseQuery, read_time: datetime.datetime,
+    ) -> None:
+        self.named_queries[name] = self._build_named_query(
```

Comment:
since this is internal, any reason to not push the datetime conversion down into this?

Reply:
Good idea. Change is made.